dbtvault-dev/macros/tables/ma_sat.sql

```diff
@@ -170,42 +170,48 @@ SELECT * FROM records_to_insert
 
 {{ dbtvault.prepend_generated_by() }}
 
-WITH source_data AS (
-    SELECT s.*
-    FROM (
-        {%- if model.config.materialized == 'vault_insert_by_rank' %}
-        SELECT {{ dbtvault.prefix(source_cols_with_rank, 'a', alias_target='source') }}
-        {%- else %}
-        SELECT {{ dbtvault.prefix(source_cols, 'a', alias_target='source') }}
-        {%- endif %}
-        ,RANK() OVER (PARTITION BY {{ dbtvault.prefix([src_pk], 'a') }}, {{ dbtvault.prefix([src_hashdiff], 'a', alias_target='source') }}, {{ dbtvault.prefix(cdk_cols, 'a') }} ORDER BY {{ dbtvault.prefix([src_ldts], 'a') }} ASC) AS source_rank
-        FROM {{ ref(source_model) }} AS a
-        WHERE {{ dbtvault.prefix([src_pk], 'a') }} IS NOT NULL
-        {%- for child_key in src_cdk %}
-            AND {{ dbtvault.multikey(child_key, 'a', condition='IS NOT NULL') }}
-        {%- endfor %}
-        ) AS s
-    WHERE s.source_rank = 1
+WITH source_data_filtered AS (
+    {%- if model.config.materialized == 'vault_insert_by_rank' %}
+    SELECT {{ dbtvault.prefix(source_cols_with_rank, 'a', alias_target='source') }}
+    {%- else %}
+    SELECT {{ dbtvault.prefix(source_cols, 'a', alias_target='source') }}
+    {%- endif %}
+    FROM {{ ref(source_model) }} AS a
+    WHERE {{ dbtvault.prefix([src_pk], 'a') }} IS NOT NULL
+    {%- for child_key in src_cdk %}
+        AND {{ dbtvault.multikey(child_key, 'a', condition='IS NOT NULL') }}
+    {%- endfor %}
     {%- if model.config.materialized == 'vault_insert_by_period' %}
         AND __PERIOD_FILTER__
     {%- elif model.config.materialized == 'vault_insert_by_rank' %}
         AND __RANK_FILTER__
     {%- endif %}
 ),
 
+{# Dedupe source data #}
+source_data AS (
+    SELECT x.*
+    FROM (
+        SELECT s.*
+            ,RANK() OVER (PARTITION BY {{ dbtvault.prefix([src_pk], 's') }}, {{ dbtvault.prefix([src_hashdiff], 's', alias_target='source') }}, {{ dbtvault.prefix(cdk_cols, 's') }} ORDER BY {{ dbtvault.prefix([src_ldts], 's') }} ASC) AS source_rank
+        FROM source_data_filtered s
+    ) AS x
+    WHERE x.source_rank = 1
+),
+
 {% if dbtvault.is_any_incremental() %}
 
 source_data_with_counts AS (
     SELECT a.*,
         ca.source_count
     FROM source_data a
-    INNER JOIN
+    CROSS APPLY
     (
         SELECT {{ dbtvault.prefix([src_pk], 't') }}, COUNT(*) AS source_count
         FROM (SELECT DISTINCT {{ dbtvault.prefix([src_pk], 's') }}, {{ dbtvault.prefix([src_hashdiff], 's', alias_target='source') }}, {{ dbtvault.prefix(cdk_cols, 's') }} FROM source_data AS s) AS t
+        WHERE {{ dbtvault.prefix([src_pk], 't') }} = {{ dbtvault.prefix([src_pk], 'a') }}
```

Comment:
Have you tested this with a composite PK? If `[src_pk]` is a list, It could cause problems

Reply:
I restored the sqlserver version of the mas code (it had been overwritten with a copy of the bigquery mas code) and did not retest it specifically. You are right in that this will not work with a composite PK, I shall correct this.

Reply:
Alec has created a DevOps ticket AB#1937 for this as it appears that this issue affects more than just mas.